### PR TITLE
Add a help panel to the inserter available in all blocks

### PR DESCRIPTION
**Pull request by youknowriad**
First step towards #16592

This expands the current "preview panel" when you hover reusable blocks in the inserter to become a help panel.

For the moment the PR only adds the "BlockCard" (title, icon and description) to that panel.
It raises some questions though:

 - Do we want to keep showing this panel on hover? How can we show it in a more permanent way but without having to click (as it would insert the block)
 - Don't mind the broken reusable blocks preview (this is broken in master too and tracked in #16808)

Capture d’écran 2019-07-30 à 2 18 04 PM

### CODE REVIEW

**Comment by kjellr** [comment]
Thanks for getting this started! 

A couple quick observations so far: 

First, the empty space in between each item causes this to be a little jarring:

dead-space

The constant show/hide gets to be a little flashy. We could fix this in one of two ways: 

1. Adjusting the hit area for the  hover, so that here are no gaps in between each item. 
2. Keep the  panel itself visible at all times, but use some sort of "Hover over a block on the left to preview it" placeholder message when a block isn't selected. In this case, at least it's just the  content that's showing/hiding, not he entire panel. 

In both cases, I think we can refine by adding some more relaxing fade animations. 

Regarding this note: 

> How can we show it in a more permanent way but without having to click (as it would insert the block)

This is a great question. I do think it would be beneficial to allow folks to keep a specific block's panel open beyond just the hover state. Especially if we place tips in there — the extra reading time is beneficial. In the current PR, this is possible by using he keyboard to activate the `:focus` state, but I'm stumped so far on a good way to enable that using just the mouse. If anyone has any great ideas, please share...

**Comment by kjellr** [comment]
> Keep the panel itself visible at all times, but use some sort of "Hover over a block on the left to preview it" placeholder message when a block isn't selected. In this case, at least it's just the content that's showing/hiding, not he entire panel.

If we go with this option (which I think is probably the simplest?) here's a first pass at a mockup for it: 

Block Library Help Panel - Default

This would provide a great place to surface the first tip from https://github.com/WordPress/gutenberg/pull/16582 as well.

**Comment by youknowriad** [comment]
@kjellr Another possibility I considered is to keep the last hovered block visible even if you move out from it.

**Comment by kjellr** [comment]
> @kjellr Another possibility I considered is to keep the last hovered block visible even if you move out from it.

🎉 I like that better. Let's try that first.

**Comment by mtias** [comment]
Keeping the last block visible could be disconcerting. I was hoping we could use this as a permanent place for saying what blocks are and showing some tips.

**Comment by kjellr** [comment]
>  Keeping the last block visible could be disconcerting. I was hoping we could use this as a permanent place for saying what blocks are and showing some tips. We could rotate different tips that don't fit in a specific block but are related to editing / insertion.

I think keeping the last block selected is worth trying, since it'll help eliminate the flash that you see when you hover over the empty areas in between the block buttons. 

The more general tips and intro panel are still definitely possible either way — It would make sense to have the panel open with general information about blocks when you first open up the inserter, and before you hover over a block icon.

**Comment by mtias** [comment]
> to have the panel open with general information about blocks when you first open up the inserter, and before you hover over a block icon.

Agreed, it'd also prevent the entire preview panel from showing / not showing which is disconcerting already.

**Comment by youknowriad** [comment]
I applied both suggestions at the same time: show a generic thing when you open the inserter (no block hovered) and avoid dismissing the last hovered block when you leave it. I can remove this if needed. Let me know how it feels.

**Comment by mtias** [comment]
I think this control is a bit excessive where it's placed.

image

Can we remove it entirely or place it in the general "settings" modal?

**Comment by mtias** [comment]
Exploring something more like this for the default view:

image

**Comment by kjellr** [comment]
> I applied both suggestions at the same time: show a generic thing when you open the inserter (no block hovered) and avoid dismissing the last hovered block when you leave it. I can remove this if needed. Let me know how it feels.

In general this interaction feels pretty good... The main thing missing is a way to get back to the original welcome panel/general message. Maybe if you mouse out of the modal, or onto the search panel, that returns? 

I suppose we could also try returning to the default panel when you hover over the right panel too, and see if that feels more natural. I may just be reacting positively to the fact that things don't flash when I'm moving the mouse in the gutter between block items, and less to the specific fact that the last item remains selected.

**Comment by youknowriad** [comment]
I moved the setting to the options modal and pushed (Github seem to have trouble to refresh the UI but the commit is there somewhere :) )

**Comment by marekhrabe** [comment]
The help panel might be a good place for more details like the one suggested in #16521. Crossposting here just in case.

**Comment by kjellr** [comment]
Pushed some tiny content + spacing changes to get us a little closer to [the mockup](https://github.com/WordPress/gutenberg/pull/16813#issuecomment-517274642): 

Screen Shot 2019-08-02 at 3 46 01 PM

**Comment by mtias** [comment]
@kjellr I created a new component for these tips. We can continue to tweak it there:

image

**Comment by kjellr** [comment]
Awesome, thanks. I'll add the copy review label so that we can get some 👀 on the wording for that default message + tip.

**Comment by mtias** [comment]
Should we try to consolidate copy review on the issue about tips in general? It'd be good to align tone and length.

**Comment by kjellr** [comment]
Yeah, we can do that in #16595. In the meantime, I think it still makes sense to have a quick review of the intro copy + headline used in that panel.

**Comment by michelleweber** [comment]
I'd delete "you can" lose the exclamation point after "more" in the second para of the description 

The "/" tip makes it sound a little like I can type just a / while writing content and it will insert a block or pull up the block search, so I'd want to make that clearer with something like "While writing in the editor, type "/" followed by a block name to quickly insert a new block." (That's how it works, right?)

**Comment by kjellr** [comment]
> I'd delete "you can" lose the exclamation point after "more" in the second para of the description

✔️ 

> The "/" tip makes it sound a little like I can type just a / while writing content and it will insert a block or pull up the block search, so I'd want to make that clearer with something like "While writing in the editor, type "/" followed by a block name to quickly insert a new block." (That's how it works, right?)

Just typing `/` does bring up a list of blocks by default, but we may need to specify that it only works in the beginning of a new paragraph. In any case, we can refine that alongside the rest of the tips in #16595 when we have our first list draft put together. I'll add a copy review tag there when it's ready. 👍 

Thanks for the help, @michelleweber!

**Comment by mapk** [comment]
I'd like to surface something for design consideration here. Tips are becoming more prevalent in the UI with #16592, #16595, and #16594. In reviewing the tips on a black background, while they're beautiful, they appear hierarchically stronger than anything else in the surrounding UI. I don't think they should because it's not the primary user action. Maybe lighter backgrounds should be considered instead.

**Comment by kjellr** [comment]
> Tips are becoming more prevalent in the UI with #16592, #16595, and #16594. In reviewing the tips on a black background, while they're beautiful, they appear hierarchically stronger than anything else in the surrounding UI.

Yeah, I do worry that a dark colored background attracts tips more attention than they deserve. When opening the block library, these tips shouldn't really be among the top tier of items you notice right away — the primary action you're meant to take in this area is to insert a block. The Tips are just auxiliary information. Their strong presence of the dark background is even more evident if we envision them spreading to other parts of the UI:

Sidebar
Frame 2 1

The tips become the first thing that you see on both those pages, and I'm not sure that's right. 

This came up in chats with @drw158 recently. He made the suggestion that we borrow from what GitHub does: the simple text and icon treatment for the ProTips that you'll find at the bottom of this page: 

Screen Shot 2019-08-07 at 8 05 41 AM

I mocked this up to see how it felt: 

Inserter-Default
Inserter-Paragraph

(I'm still not sold on what color the icon should be, but) I do like this exploration. This is simple enough that it expands to the other use cases quite well. The tips are present, but secondary to other actions on the screen: 

Sidebar
Frame 2

It's also worth noting that the dark-colored tips may still have a good use case, even if we went with this lighter treatment here. If we implement a guided tour NUX experience as discussed in https://github.com/WordPress/gutenberg/issues/16315, it would make sense sense for those to be treated more prominently. Those tips would be opted into, and would be walking the user through the experience. In that case we _would_ want the tips to be the first thing the users sees on each page they appear. 

Interested in hearing further thoughts from the group. Either way, we'll have time and space to iterate on this later.

**Comment by mtias** [comment]
@kjellr can you try it without the divider line? Also feel free to push it to the component, we can see how it feels in practice.

**Comment by kjellr** [comment]
Just pushed this up: 

Screen Shot 2019-08-07 at 2 46 28 PM

**Comment by mapk** [comment]
Looking good! I'm digging it, even without the divider. It feels lighter, doesn't grab me right away, and the positioning is steller. I really like this positioning extending to the Block Setting sidebar as well. Thanks @kjellr!

**Comment by youknowriad** [comment]
Rebased that branch. the conflicts were not simple so let me know if there's any breakage.

**Comment by mtias** [comment]
Nice! I like it a fair bit more without the divider line. Let's get it merged and see how it scales to other places :)

**Comment by karmatosed** [comment]
I would add a +1 for divider but that's an easy enough fix. I think it will help visually to define but not stand out.

**Comment by youknowriad** [comment]
I think this PR will show its value once we add the example API to preview any block in the inserter cc @retrofox @marekhrabe 

Anyway, it should be ready for a final review now.

**Comment by shaunandrews** [comment]
image

Should the right panel's shadow appear on-top of the left panel? It feels weird, and if we're going by the Material understanding of Elevation, this implies that the right panel is "higher up" than the left — when I think they should be at the same height.

**Comment by kjellr** [comment]
> Should the right panel's shadow appear on-top of the left panel? It feels weird, and if we're going by the Material understanding of Elevation, this implies that the right panel is "higher up" than the left — when I think they should be at the same height.

Yeah, I tend to agree. This just borrows from what's live already for the block styles preview panels: 

Screen Shot 2019-08-08 at 11 29 52 AM

... but I think it's more of a symptom of the structure than an intentional choice. Ideally there should be a wrapper around the entire container, with a single box shadow assigned to that. But at the moment, each panel gets its own box shadow.

**Comment by kjellr** [comment]
@youknowriad, Returning to the earlier part of the thread regarding the hover state: Would it be possible to switch back to the default help panel state if a user hovers over the panel headers, or over the search bar? That might be a relatively natural way to effectively "deselect" a block and return to the introductory panel. 

If it's too tough, I think we can push to a further PR, but I think it'd be great to allow for people to get that default message back if they mistakenly hovered over a block while reading it.

**Comment by simison** [comment]
How would this look like on small screens like mobile phones? Should it be hidden, or should there be a separate button for each block to bring it visible?

**Comment by youknowriad** [comment]
It's hidden on mobile screens right now. I think that's fine because this is additional information.

**Comment by youknowriad** [comment]
@kjellr I restored the original hover behavior because I noticed issues, the block insertion point staying visible after closing the inserter in some occasions. Maybe we can start this way and iterate on it now that we have the "default state"

**Comment by kjellr** [comment]
> Maybe we can start this way and iterate on it now that we have the "default state"

Sounds good. 👍

**Comment by retrofox** [comment]
This is a tiny detail, but when I was testing saw a link in the popover, 

image

and wasn't able to click on it, which leads me to wonder if we are going to make this help pane more interactive in the near future. If so, maybe we should consider a way to keep the pane there.

**Comment by kjellr** [comment]
> when I was testing saw a link in the popover,

Ah, good point. I forgot that some third party blocks include links in those. Just a couple hours ago you would've been able to click that, but we [removed](https://github.com/WordPress/gutenberg/pull/16813#issuecomment-519869747) the "stay open after hover" functionality since then. 🤔 

If we don't get that sorted out by this first merge, I wonder if it's possible to strip those out for v1? I think it's probably fine to highlight those support links only _after_ someone's inserted the block.

**Comment by mtias** [comment]
> If we don't get that sorted out by this first merge, I wonder if it's possible to strip those out for v1? I think it's probably fine to highlight those support links only after someone's inserted the block.

This is out of control for now, since it comes from the block description and we cannot know what is being added there by block developers.

**Comment by kjellr** [comment]
> This is out of control for now, since it comes from the block description and we cannot know what is being added there by block developers.

Ah, ok. I'd thought that #11330 had been merged already, which would standardize that a bit.

**Comment by kjellr** [comment]
I pushed a small update that adds a quick fade-in animation to the content of the panel. This cuts down slightly on the flash of default panel content when you quickly hover laterally through items. The flash is still visible depending on how fast you move your mouse though, so it's far from a perfect fix: 

fade

File: packages/block-editor/src/components/inserter/style.scss [hunk 1]
```diff
@@ -161,6 +161,7 @@ $block-inserter-search-height: 38px;
 		padding-bottom: 20px;
 		margin-bottom: 10px;
 		border-bottom: $border-width solid $light-gray-500;
+		@include edit-post__fade-in-animation();
```
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/style.scss]: Why is this animation called `edit-post` ? Is this in the global mixin, should we give it a generic name?

**Comment by senadir** [comment]
I think this PR is in a stable state to be merged right? with other PRs regarding preview & switcher component being merged, I see no reason for this to be blocked, as @mtias said, the link issue is already too late, using any other behavior for the hover will the break the initial state, so I feel like this is best current solution.

I've tested with with some themes on safari, firefox, chrome in OSX and it works fine
inserter-help

I approved the PR but if anyone feels otherwise, then it could be stopped

**Review by senadir — APPROVED** [review]

**Comment by mapk** [comment]
I tested this again (in FF) to see if I can recreate Kjell's issue above. Turns out I'm getting a rapid appearance of the default tips info between each of the block buttons. First, this is completely normal because of the spacing in between each block button, but it doesn't feel groovy. 

help

Can we make it so the block buttons are adjacent with one another in the Inserter? Removing the padding from `.block-editor-block-types-list__list-item` will fix this.

fixed

**Review by mapk — CHANGES_REQUESTED** [review]
As mentioned in my previous comment, I've got a change request.

Should we remove the padding from the Inserter's block buttons here: `.block-editor-block-types-list__list-item`?

**Comment by youknowriad** [comment]
I think that padding is important once we have a "parent/children" block (like the "contact form" block from Jetpack)

**Comment by youknowriad** [comment]
Also sometimes the block label height is not the same across blocks so we end up with space anyway.

**Comment by mapk** [comment]
> I think that padding is important once we have a "parent/children" block (like the "contact form" block from Jetpack)

Should we rethink this? It doesn't seem like Core uses this pattern for any of our "parent" blocks like Media+Text block, Columns block, Group block. Is this because they don't really have "children" but rather just "nested" blocks? Do we need to differentiate this to the user?

**Comment by youknowriad** [comment]
> Do we need to differentiate this to the user?

I think these are two different things. The idea is that some blocks can only be inserted as children of a given parent block. There's a strong relationship between the children and the parent. I think the "Menu" block is probably something like that while it's different for Media & Text blocks.

We could rethink the design of these blocks in the inserter, it's not entirely clear whether the distinction made today is necessary or not.

**Comment by youknowriad** [comment]
Now that #17083 is merged, we can consider removing the space between the blocks in the inserter to unblock this. I'm going to try it.

**Comment by youknowriad** [comment]
@mapk I removed the horizontal space, let me know what you think?

File: packages/block-editor/src/components/inserter/style.scss (hunk 1 above)
Comment (kjellr) [reply packages/block-editor/src/components/inserter/style.scss]: Yeah, it should — we use it in many different places. I'll open a separate PR for that.

**Comment by mapk** [comment]
This looks great now! It's working smoothly on the horizontal cursor flow between block buttons. 

👍 :shipit:

**Review by mapk — APPROVED** [review]

**Comment by youknowriad** [comment]
Let's try it thanks everyone.

**Comment by rickybanister** [comment]
It might be nice to use the help panel to display the block description for the hovered or highlighted block, providing more context for the element you're considering adding to the page. 

Rough mockup:

Screen Shot 2019-08-22 at 9 29 48 AM

**Comment by senadir** [comment]
@rickybanister this is the current behavior, pull master to the latest version to see it

pic from master
image

**Comment by rickybanister** [comment]
haha, sorry for not testing, great work everyone!

**Comment by talldan** [comment]
Noted that this PR removed some padding around the blocks in the menu in order to keep the transitions in the side panel smooth. It doesn't seemed to have stopped the issue when moving the mouse up and down.

An alternative might be to keep the spacing as it was but use a small delay on the side panel item disappearing.

**Comment by CreativeDive** [comment]
I'm just wondering what to do with this big empty space?

empty-area

This area is perfect for displaying a reusable blocks selection for each block type. This would be great if we could select the most used reusable block directly from this panel.

reusable-blocks

This would be awesome and encourage the user to work even more with reusable blocks :-)

**Comment by CreativeDive** [comment]
@kjellr Thanks for reply. I think a reusable block selection would be more interesting and useful :-)

**Pull request merged**